File: frontend/cxx/evutils.cxx
#include "evutils.h"
#include <stdint.h>

unsigned short int GetEC(RawEvent &event) {
   return(UnmaskEC(event[1]));
}

unsigned short int GetLEN(RawEvent &event) {
   return(UnmaskLEN(event[event.size() - 2]));
}

unsigned short int GetCRC(RawEvent &event) {
   return(UnmaskCRC(event[event.size() - 1]));
}

std::vector<unsigned short int> GetChannelSamples(RawEvent &ev, unsigned short int m, unsigned short int ch, unsigned short ns) {
   std::vector<unsigned short int> chdata;
   int offset = 4;
   int start = m*(32+2+(ns*32)) + ch*(ns+1) + offset;
   int end = start + ns;
   for(int i=start; i<end; i++)
      chdata.push_back(ev[i]);
   return(chdata);
}

unsigned int GetTime(RawEvent &event) {
   int msb = event[event.size() - 4] << 16;
   int lsb = event[event.size() - 3] << 1;
   return (msb + lsb) >> 1;
}

unsigned short int UnmaskEC(unsigned short int word) {
   return (word & 0x0FFF);
}

unsigned short int UnmaskLEN(unsigned short int word) {
   return (word & 0x0FFF);
}

unsigned short int UnmaskCRC(unsigned short int word) {
   return (word & 0x00FF);
}

bool CheckEC(RawEvent &event, unsigned short ns) {
   uint16_t i = 1;
   uint16_t readec = UnmaskEC(event[i]);
   int step = 2 + 32 + (32 * ns);
   while (i < event.size()) {
      if(UnmaskEC(event[i]) != readec) {
         return false;
      }
      i += step;
   }
   return true;
}

bool CheckLEN(RawEvent &event) {
   uint16_t evlen = GetLEN(event);
   int length = event.size() + 1;
   return (evlen == length);
}

bool CheckCRC(RawEvent &event) {
   uint16_t evcrc = GetCRC(event);
   int crc = 0;
   for(uint16_t  i=0; i<event.size()-1; i++) {
      uint8_t msb = (event[i] & 0xFF00) >> 8;
      uint8_t lsb = (event[i] & 0x00FF);
      crc = (crc ^ (msb ^ lsb));
   }
   uint8_t crc8 = crc & 0x00FF;
   return (evcrc == crc8);
}

File: utils/dumpevents/lib/evutils.py

class EventUtils():

    def __init__(self):
        self.debug = False
        self.samplenum = 1
        self.modulenum = 8

    def SetDebug(self, value):
        self.debug = value

    def SetModuleNum(self, value):
        self.modulenum = value

    def GetModuleNum(self):
        return self.modulenum

    def SetSampleNum(self, value):
        self.samplenum = value

    def GetSampleNum(self):
        return self.samplenum

    def PrintTab(self, ev):
        col = 0
        num = 0
        print('%04d-> ' % num, end='')
        for item in ev:
            num = num + 1
            print('0x%04X' % item, end='')
            col = col + 1
            if(col == 16):
                col = 0
                print('\n')
                print('%04d-> ' % num, end='')
            else:
                print(' ', end='')
        print('\n')

    def PrintCol(self, ev):
        for item in ev:
            print('%X' % item)
        print('\n')

    def UnmaskEC(self, word):
        return (word & 0x0FFF)

    def GetEC(self, ev):
        return (self.UnmaskEC(ev[1]))

    def UnmaskLEN(self, word):
        return (word & 0x0FFF)

    def GetLEN(self, ev):
        return (self.UnmaskLEN(ev[-2]))

    def UnmaskCRC(self, word):
        return (word & 0x00FF)

    def GetCRC(self, ev):
        return (self.UnmaskCRC(ev[-1]))

    def CheckEC(self, ev):
        ec = 0
        i = 1
        readec = self.UnmaskEC(ev[i])
        step = 2 + 32 + (32 * self.samplenum)
        while (i < len(ev)):
            if (self.UnmaskEC(ev[i]) != readec):
                if self.debug:
                    print("Event EC: %X != %X not consistent" % (self.UnmaskEC(ev[i]), readec))
                return False
            i += step
        return True

    def CheckCRC(self, ev):
        evcrc = self.GetCRC(ev)
        crc = 0
        for item in ev[:-1]:
            msb = (item & 0xFF00) >> 8
            lsb = (item & 0x00FF)
            crc = (crc ^ (msb ^ lsb))
        crc8 = crc & 0x00FF
        if self.debug:
            print("Event CRC: %X - Calculated CRC: %X" % (evcrc, crc8))
        return (evcrc == crc8)

    def CheckLength(self, ev):
        evlen = self.GetLEN(ev)
        length = len(ev) + 1
        if self.debug:
            print("Event LEN: %d - Calculated LEN: %d" % (evlen, length))
        return (evlen == length)

    def GetChannelSamples(self, ev, m, ch):
        offset = 4
        start = m*(32+2+(self.samplenum*32)) + ch*(self.samplenum+1) + offset
        end = start + self.samplenum
        return(ev[start:end])

    def GetTime(self, ev):
        msb = ev[-4]
        lsb = ev[-3]
        msb = msb << 16
        lsb = lsb << 1
        time = (msb + lsb) >> 1
        return time
